src/app/util/attribute-storage.h

```diff
@@ -201,6 +201,12 @@ const EmberAfCluster * emberAfGetClusterByIndex(chip::EndpointId endpoint, uint8
 //
 chip::Span<const EmberAfDeviceType> emberAfDeviceTypeListFromEndpoint(chip::EndpointId endpoint, CHIP_ERROR & err);
 
+//
+// Retrieve the tag list associated with a specific endpoint.
```

Comment:
Ownership is not specified. A span is non-owning. This has to be made clear throughout, since that means that dynamic tag lists may be a problem across calls. Returning spans of complex types should be avoided if possible. Perhaps a `GetTagListItemFromEndpoint` and `GetTagListSizeFromEndpoint` would allow a safer access, with storage for the `chip::app::Clusters::Descriptor::Structs::SemanticTagStruct::Type` provided by the caller in the method, so that it's filled by the method if there. 

Otherwise, an iterator may be better: see src/lib/support/CommonIterator.h  and GroupDataProvider.h for an example of usage. That interface is made specifically for this type of usecase.

Reply:
Overall, I get that this is hard to possibly fix before SVE, but this is a brand new global public API that will be very hard to refactor later, so this should have come 2 months ago, not just before SVE...

Reply:
Thanks, I think we can implement this in two steps,
step 1, we should make sure implement  api `CHIP_ERROR GetTagListFromEndpointAtIndex(chip::EndpointId endpoint, size_t index, chip::app::Clusters::Descriptor::Structs::SemanticTagStruct::Type & tag);`   before SVE (SDK freeze time , Aug 25th).
step 2, we implement a class to save the SemanticTagStruct provide to caller that doesn't need to care the span. (During SVE and before 1.2 release).